File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int left = 0;
        int right = height.length-1;
        int lMax = height[left];
        int rMax = height[right];
        int water = 0;

        while(left < right){
            if(lMax < rMax){
                left++;
                lMax = Math.max(height[left], lMax);
                water += lMax - height[left];
            }
            else{
                right--;
                rMax = Math.max(height[right], rMax);
                water += rMax - height[right];
            }
        }

        return water;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        lMax = height[left]
        rMax = height[right]
        water = 0

        while (left < right):
            if (lMax < rMax):
                left += 1
                lMax = max(lMax, height[left])
                water += (lMax - height[left])
            else:
                right -= 1
                rMax = max(rMax, height[right])
                water += (rMax - height[right])
        
        return water